ssz/sedes/list.py

```diff
@@ -42,80 +59,53 @@ def __init__(self, element_sedes=None, empty=False):
-                val
-            )
-        serialized_len = len(serialized_iterable_string).to_bytes(LIST_PREFIX_LENGTH, 'big')
-
-        return serialized_len + serialized_iterable_string
-
-    def deserialize_segment(self, data, start_index):
-        """
-        Deserialize the data from the given start_index
-        """
-        # Make sure we have sufficient data for inferring length of list
-        if len(data) < start_index + LIST_PREFIX_LENGTH:
-            raise DeserializationError(
-                'Insufficient data: Cannot retrieve the length of list',
-                data
+                "Can only serialize empty Iterables"
             )
 
-        # Number of bytes of only the list data, excluding the prepended list length
-        list_length = int.from_bytes(data[start_index:start_index + LIST_PREFIX_LENGTH], 'big')
-        list_end_index = start_index + LIST_PREFIX_LENGTH + list_length
-        # Make sure we have sufficent data for inferring the whole list
-        if len(data) < list_end_index:
-            raise DeserializationError(
-                'Insufficient data: Cannot retrieve the whole list data',
-                data
-            )
+    @to_tuple
+    def deserialize_content(self, content: bytes) -> Generator[S, None, None]:
+        if self.empty and len(content) > 0:
+            raise DeserializationError(f"Serialized list is not empty")
 
-        deserialized_list = []
-        # element_start_index is the start index of an element in the serialized bytes string
-        element_start_index = start_index + LIST_PREFIX_LENGTH
-        while element_start_index < list_end_index:
-            element, element_start_index = self.element_sedes.deserialize_segment(
-                data, element_start_index
+        element_start_index = 0
+        while element_start_index < len(content):
+            element, next_element_start_index = self.element_sedes.deserialize_segment(
+                content,
+                element_start_index,
             )
-            deserialized_list.append(element)
 
-        return tuple(deserialized_list), list_end_index
+            if next_element_start_index <= element_start_index:
+                raise Exception("Invariant: must always make progress")
+            element_start_index = next_element_start_index
 
-    def deserialize(self, data):
-        deserialized_data, end_index = self.deserialize_segment(data, 0)
-        if end_index != len(data):
-            raise DeserializationError(
-                'Data to be deserialized is too long',
-                data
-            )
+            yield element
 
-        return deserialized_data
+        if element_start_index > len(content):
+            raise Exception("Invariant: must not consume more data than available")
 
 
 address_list = List(address)
 boolean_list = List(boolean)
 bytes_list = List(bytes_sedes)
-empty_list = List(empty=True)
+empty_list: List[None, None] = List(empty=True)
```

Comment:
Are `None, None` the proper types here? I would have thought they'd still be something like `Sequence` or `Iterable` types but maybe I'm not understanding something.

Reply:
The type hint is for the element, not for the whole list (as lists in SSZ are required to have the same type for each element). The empty list has no element, so it shouldn't matter, therefore I just picked `None`.